rescompy/rescompy.py

```diff
@@ -815,98 +956,518 @@ def train(
+            
+            if (num_batches > 0 and remainder_length):
+                start_time = transient_i + num_batches * batch_size
+                inputs_i = inputs[task_ind][start_time:]
+                states_i = self.get_states(new_state, inputs_i)
+                new_state = states_i[-1]
+                states_i = np.concatenate((lookback_states, states_i[1:]))
+                inputs_i = np.concatenate((lookback_inputs, inputs_i[1:]))
+                lookback_inputs = inputs_i[-lookback_length-1:]
+                lookback_states = states_i[-lookback_length-1:]
+                
+                targets_i = target_outputs[task_ind][start_time:]
+                features_i = feature_function(states_i, inputs_i)
+
+                SS_T += features_i.T @ features_i
+                VS_T += features_i.T @ targets_i
+
+        if (num_batches > 1 or (num_batches == 1 and remainder_length)):
+            inputs_saved = [inputs_i[lookback_length:]]
+            states_saved = [states_i[lookback_length:]]
+        else:
+            inputs_saved = inputs_i
+            states_saved = states_i
+        targets_saved = [targets_i]
+        transients_saved = [transient_i]
+
+        # use the signature of the regression to figure out the parameters
+        params = inspect.signature(regression).parameters
+        # assume u = input, s = feature, v = target, g is the feature function
+        map = {
+		  'SS_T': SS_T,
+		  'VS_T': VS_T
+        }
+        # check that there are no unknown parameters
+        if any([k not in map for k in params.keys()]):
+            logging.error('Please ensure regression function signature is valid')
+        # construct the parameter dict 
+        args = {arg: map[arg] for arg in params.keys()}
+
+        # Optimize the output weights      
+        weights = regression(**args)
+        
+        # Construct and return the training result.
+        return TrainResult(states_saved, inputs_saved, targets_saved,
+						   transients_saved, [-1],
+						   feature_function, weights)
+
+
+    @validate_arguments(config=dict(arbitrary_types_allowed=True))
+    def train_batched(
+        self,
+        transient_lengths: Union[int, List[int]],
+        inputs:            Union[np.ndarray, List[np.ndarray]],
+        target_outputs:    Union[np.ndarray, List[np.ndarray], None] = None,
+        initial_state:     Optional[np.ndarray]                      = None,
+        dg_du:             Optional[np.ndarray]                      = None,
+        feature_function:  Union[Callable, features.ESNFeature]      = features.StatesOnly(),
+        regression:        Optional[Callable]                        = regressions.batched_ridge(),
+        batch_size:        int                                       = 10,
+        accessible_drives: Union[int, List[int], str]                = "all",
+        ) -> TrainResult:
+        """The training method.
+                
```

Comment:
what is the difference between train and train batched?  maybe add some text here to the comment

Reply:
The batch training method is for use with regression functions that accept the regression matrices SS_T and VS_T as arguments. These matrices are incremented every batch_size training signals to save memory.

The "standard" train routine is for use with general regression functions that accept a combination of the inputs (u), features (s), targets (v), feature_function (g), or Jacobian (dg_du) as argument. The features, inputs, and targets over the entire training period are stored until training has finished.

Now that the signature of the regression function is checked before calling it, I may be able to merge them into the same function.However, I'm not sure how easily that could be done without sacrificing some of the memory that is saved in the current configuration.

Reply:
I've added some text to the comments.